pingcap/tidb-operator
# Support smooth migration

### What problem does this PR solve? 
https://github.com/pingcap/tidb-operator/issues/2895
### What is changed and how does it work?
1. Add `spec.PDAddress` of `[]string` to TidbCluster CR
2. The format of `spec.PDAddress` looks like `http://{address}:{port}`. `https` portocal is not supported at present.
3. When creating a tidb cluster with`spec.PDAddress` specified, PDs will join the exsiting PD cluster, which address is `spec.PDAddress`.
### Check List 

Tests 

 - Unit test
 - Manual test (add detailed scripts or steps below)
   1. Creating a tidb cluster in old version `tidb-operator`, and create 8 tables, each 1000000 rows.
   2. Upgrading `tidb-operator` to new version with `spec.PDAddress`. The existing tidb cluster is running without influenced. 
   3. Creating a tidb cluster with `spec.PDAddress` specified(If the format not matchs regex of `^http://.*:[0-9]+$`, it won't be created).
   4. Transfering PD leaders to the new tidb cluster, and scaling in PD replicas to 0. It can still work normally.
   5. Deleting all the stores in the old tidb cluster, and then deleting the old cluster.
   6. Notice that all the datas are transfered successfully.
   7. Removing the `spec.PDAddress` or not has no effect to the cluster.


Code changes

 - Has Go code change


Related changes

 - Need to update the documentation

### Does this PR introduce a user-facing change?:

```release-note
NONE
```
